### DISCUSSION TITLE
Refreshing the database connection

### DISCUSSION THREAD

**Discussion by BananaMasterz** [simolus3/drift#2862] [Q&A]
Sometimes the connection to my database is lots (for 3rd party reasons) and I want to close and reopen my connection.
I'm using provider and I'm not sure how to approach this... here's the current code:
```dart
class MyDatabase extends _$MyDatabase with m.ChangeNotifier {
  MyDatabase() : super(_openConnection());
  void refreshConnection() async {
    // close and reopen here
  }
  @override
  int get schemaVersion => 1;
}
```
when I do `db.close()` after loosing internet connection for example and regaining it, the function hangs and is awaited forever trying to close the db. No error is thrown either. Is there a way to know where the error happens? Cuz no message so can't know.

**Update** for the hanging `db.close()`
in postgres-3.0.7 package in connection.dart I found it's getting stuck here forever: 

```dart
Future<void> _close(bool interruptRunning, PgException? cause) async {
    if (!_isClosing) {
      _isClosing = true;

      print('s1');
      if (interruptRunning) {
        _pending?.handleConnectionClosed(cause);
        _channel.sink.add(const TerminateMessage());
      } else {
        // Wait for the previous operation to complete by using the lock
        print('s2');
        await _operationLock.withResource(() {
          // Use lock to await earlier operations
          _channel.sink.add(const TerminateMessage());
        });
        print('s4');
      }

      await Future.wait([_channel.sink.close(), _serverMessages.cancel()]);
    }
  }
  ```
  print s4 is never printed. If I remove this lock then it works normally. This is the other packages problem so don't mind it. I'll probably find something. Only interested in the listener now that lets me act upon loss of connection

**Comment by BananaMasterz** [comment]
more specifically I want to have a wrapper around the db class that whenever a query fails due to closed connection (or if  it's possible to know when the connection is broken), it retries to connect (don't mind the retry logic and when it should stop trying I can manage that)

**Comment by simolus3** [comment]
> This is the other packages problem so don't mind it

Well, you'll never guess who wrote that snippet :grimacing:  But I agree that opening an issue on the `postgres` package is the right way to go about this, I can still take a look at the issue then.

> more specifically I want to have a wrapper around the db class that whenever a query fails due to closed connection

It's probably easier to not do this around the whole database class, but rather over the underlying connection only. Then you don't have to open the whole database class again. You can take a look at the [postgres](https://github.com/simolus3/drift/blob/develop/extras/drift_postgres/lib/src/pg_database.dart) implementation for this and possibly adapt it to handle connection errors. But it's probably very tricky to get right, I imagine handling transactions might be a challenge.

**Reply by BananaMasterz** [reply]
> Well, you'll never guess who wrote that snippet 😬

oh... then maybe u know better what are the implications of me removing it? 🤷‍♂️

As for the rest, I don't think I care about transactions would be an issue since if the connection fails everything breaks. If everything is broken there's nothing to worry about (right?)

**Reply by simolus3** [reply]
> oh... then maybe u know better what are the implications of me removing it? 🤷‍♂️

A slightly safer change might be to change `close()` to call `_close()` with `interruptRunning: true`.

> I don't think I care about transactions would be an issue since if the connection fails everything breaks

Fair, my point is that you still need to surface an exception for transactions. Outside of transactions, you can just recreate the operation on a new connection if the old one is gone.
